lib/compress/zstd_compress.c

```diff
@@ -4556,7 +4584,7 @@ static size_t ZSTD_compress_frameChunk(ZSTD_CCtx* cctx,
                 }
             }  /* if (ZSTD_useTargetCBlockSize(&cctx->appliedParams))*/
 
-
+            if (cSize < blockSize) savings += (blockSize - cSize);
```

Comment:
Do we also need to subtract savings if we grow the block?

If we aren't precisely tracking the `savings` I can see cases where an adversarial pattern could cause compression to fail.

Reply:
OK, we can add some malus when a block is not compressed.
This will be a conservative over-estimate, but that's not important, we just want to make sure there is no possible scenario that could expand data by more than 3-bytes per full block.

Reply:
> OK, we can add some malus when a block is not compressed.

I don't like this solution because it is a patch that hides the problem. But I don't see how it guarantees that it can never occur. I'd prefer to precisely calculate the `savings` so we can know exactly if we are allowed to split or not.

Reply:
I've added a paragraph which explains why a 1-byte malus to `savings` for each incompressible block is enough to guarantee that the 3-bytes per 128 KB expansion limit cannot be breached.